MSAL/src/native_auth/cache/MSALNativeAuthCacheAccessor.swift

```diff
@@ -0,0 +1,181 @@
+            let idToken = try tokenCacheAccessor.getIDToken(
+                forAccount: account.lookupAccountIdentifier,
+                configuration: accountConfiguration,
+                idTokenType: MSIDCredentialType.MSIDIDTokenType,
+                context: context)
+            let refreshToken = try tokenCacheAccessor.getRefreshToken(
+                withAccount: account.lookupAccountIdentifier,
+                familyId: nil,
+                configuration: accountConfiguration,
+                context: context)
+            let accessToken = try tokenCacheAccessor.getAccessToken(
+                forAccount: account.lookupAccountIdentifier,
+                configuration: accountConfiguration,
+                context: context)
+            return MSALNativeAuthTokens(accessToken: accessToken, refreshToken: refreshToken, rawIdToken: idToken.rawIdToken)
+        }
+
+    func getAccount(
+        accountIdentifier: MSIDAccountIdentifier,
+        authority: MSIDAuthority,
+        context: MSIDRequestContext) throws -> MSIDAccount? {
+            return try tokenCacheAccessor.getAccountFor(
+                accountIdentifier,
+                authority: authority,
+                realmHint: nil,
+                context: context)
+        }
+
+    func getAllAccounts(configuration: MSIDConfiguration) throws -> [MSALAccount] {
+        let request = MSALAccountsProvider(tokenCache: tokenCacheAccessor,
+                                           accountMetadataCache: accountMetadataCache,
+                                           clientId: configuration.clientId,
+                                           externalAccountProvider: externalAccountProvider)
+        return try request?.allAccounts() ?? []
+    }
+
+    func validateAndSaveTokensAndAccount(
+        tokenResponse: MSIDTokenResponse,
+        configuration: MSIDConfiguration,
+        context: MSIDRequestContext) throws -> MSIDTokenResult? {
+            let ciamOauth2Provider = getCIAMOauth2Provider(clientId: configuration.clientId)
+            return try? validator.validateAndSave(tokenResponse,
+                                                  oauthFactory: ciamOauth2Provider.msidOauth2Factory,
+                                                  tokenCache: tokenCacheAccessor,
+                                                  accountMetadataCache: accountMetadataCache,
+                                                  requestParameters: getRequestParameters(tokenResponse: tokenResponse,
+                                                                                          configuration: configuration,
+                                                                                          context: context),
+                                                  saveSSOStateOnly: false)
+        }
+
+    // Here we create the MSIDRequestParameters required by the validateAndSave method
+    private func getRequestParameters(
```

Comment:
why do you need this method? MSIDRequestParameters should be created in the top layer api (when dev calls our public api) and should contain all params provided from the dev. MSIDConfiguration is also part of MSIDRequestParameters.

Reply:
On our end, the only case where we needed a `MSIDRequestParameters` instance was when we accessed the cache. The reason why we don't use an instance of `MSIDRequestParameters` from the top layer is because for us a lot of the parameters needed by `MSIDRequestParameters` are not present. For example, one of the main ones is the `redirectURL` which we don't need as we make all calls directly to the server, without the use of the UIWebView use by Auth UX.
Another thing is that we don't need the scopes all the time. For example when we initiate the process of sign up of an user we don't pass in the scopes:
```
public func signUpUsingPassword(
        username: String,
        password: String,
        attributes: [String: Any]? = nil,
        correlationId: UUID? = nil,
        delegate: SignUpPasswordStartDelegate
    )
```
We only pass those when we need to call the token endpoint do do a Sign In operation. One other case where we don't use scopes is the Reset Password operation as they are not needed.
Also because of how CIAM/Microsoft Entra is designed we are missing stuff like `providedAuthority` or `cloudAuthority`
As for the function itself, why it's needed it's used to interact with the `MSIDTokenResponseValidator` class (specifically the `validateAndSaveTokenResponse` function) which doesn't necessarily require the "network" side of the `MSIDRequestParameters` but rather the telemetry and properties that are used to interact with the cache and save the token.
Up to this point, we don't need to create an instance of `MSIDRequestParameters` and this is the only place in our code where we use it and we build it based on the data we pass around - `MSIDConfiguration` and `MSIDRequestContext`